### ISSUE TITLE
macOS support for being a subnet router ("relay node") or exit node

### ISSUE THREAD

**Issue by bilby91** [tailscale/tailscale#911]
**Is your feature request related to a problem? Please describe.**

I would like to be able to add subnet routes from a macOS client. I run some VPN connections on my macOS client and would like that other tailscale clients be able to route the traffic over by macOS machine.

**Describe the solution you'd like**

The solution would be to enable the functionality on macOS clients, currently only supported on linux clients.

**Describe alternatives you've considered**

None

**Additional context**

None

<img src="https://frontapp.com/assets/img/favicons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_399u9)

**Comment by bradfitz** [comment]
Well, the bigger issue is our official Mac build can only run as a user, not e.g. at system boot in the background.

So if your Mac reboots, Tailscale is down.

You can fix that by running the open source tailscaled under launchd, and at that point I think the community even has this subnet router stuff working (@danderson? I seem to recall code for this on darwin).

So #718 for being a System Extension is interesting, and #177 (Homebrew port) is also interesting, which I see has stuff like:

```
$ brew services |grep tailscale
tailscale  started root /Library/LaunchDaemons/homebrew.mxcl.tailscale.plist
```

But I'll keep this bug open as the meta bug for the high level request, which I agree should exist somehow.

**Comment by bilby91** [comment]
Thanks for the response @bradfitz !

I tried installing tailscale from source. I don't see the subnet options using the help command:

```
╰─ sudo ~/go/bin/tailscaled --help
Usage of /Users/bilby91/go/bin/tailscaled:
  -cleanup
    	clean up system state and exit
  -debug string
    	listen address ([ip]:port) of optional debug server
  -fake
    	use userspace fake tunnel+routing instead of kernel TUN interface
  -port value
    	UDP port to listen on for WireGuard and peer-to-peer traffic; 0 means automatically select (default 0)
  -socket string
    	path of the service unix socket (default "/var/run/tailscale/tailscaled.sock")
  -state string
    	path of state file
  -tun string
    	tunnel interface name (default "tailscale0")
  -version
    	print version information and exit
```

Is there any ETA for the brew version ?

**Comment by bilby91** [comment]
@bradfitz Let me take that back :)

I was able to run `tailscaled` with `sudo` and then call `tailscale up -advertise-routes 10.4.0.0/18`. I also enabled the subnet routes via the admin panel. I still can't connect to a resource under the `10.4.0.0/18` subnet from a different tailscale node.

 The `10.4.0.0/18` subnet is accessible from my mac via a vpn connection that I'm running. I'm missing something or this should be working in theory ?

**Comment by bradfitz** [comment]
I'm not up to date on the state of the open source Mac build. The folks in #177 might know.

Does it work at all? That is, what's your tailscale 100.x.y.z IP? Can you ping that at least from another machine running Tailscale? Can you ping `hello.ipn.dev` from the mac?

The `--tun` value might be wrong, too? I think it might need be `utun2` or something on darwin, but that's a hazy memory.

**Comment by bilby91** [comment]
Yes, it's working! I can ssh into my boxes while running the `tailscaled` daemon myself. I had to use `--tun utun4` if I recall correctly.

**Comment by bradfitz** [comment]
I think this is all working now.

**Comment by bilby91** [comment]
@bradfitz Thanks! 

I can confirm that my macOS client can now route into advertised networks from a Linux node. Is there any plan for adding advertised networks on the macOS client app? I want to basically do the opposite, advertise a set of networks from my macOS client so that the Linux node can route into them.

Thanks!

**Comment by bradfitz** [comment]
Reopening, as I was wrong before. There just isn't code for this.

From my dup bug I filed earlier:

---

tailscaled runs reasonable well on macOS (https://github.com/tailscale/tailscale/wiki/Tailscaled-on-macOS), but there are a few missing features compared to the Linux version. One is being a subnet router ("relay node").

This bug tracks adding subnet routing to `tailscaled`-on-macOS. (This is **NOT** about the GUI version of Tailscale on macOS)

Related tracking bug for Windows: https://github.com/tailscale/tailscale/issues/835

One option is doing it purely in userspace: https://github.com/tailscale/tailscale/issues/707

But this bug is specifically about having the macOS kernel do the work, using `pfctl` or whatnot:

* https://apple.stackexchange.com/questions/363099/how-to-forward-traffic-from-one-machine-to-another-with-pfctl
* https://serverfault.com/questions/791181/redirecting-traffic-to-a-specific-address-and-port-using-pf-on-macos
* etc

**Comment by bradfitz** [comment]
Decision from meeting: we'll make the macOS Network/SystemExtension mode also support being a subnet router (& exit node) via the CLI in hybrid netstack mode, for Tailscale 1.12.

**Comment by agrw** [comment]
So does the macOS Network/SystemExtension mode support being a subnet router (& exit node) via the CLI in hybrid netstack mode in Tailscale 1.12? or will macOS have ability to be an exit node in next version?

**Comment by jdelman** [comment]
> Decision from meeting: we'll make the macOS Network/SystemExtension mode also support being a subnet router (& exit node) via the CLI in hybrid netstack mode, for Tailscale 1.12.

It seems like maybe this was bumped from the 1.12.0 release. Are there still plans to include this functionality in the macOS build of Tailscale at all?

**Comment by DentonGentry** [comment]
We do still intend to support MacOS as an exit node. This issue tracks work toward that goal.

In the 1.12 release we fixed https://github.com/tailscale/tailscale/issues/2302, which was a blocking issue for MacOS subnet routing and exit node support as it uses the same userspace netstack. We're on a monthly release cadence now, the next opportunity to deliver MacOS support is 1.14 in about a month.

**Comment by jdelman** [comment]
Thanks for the update @DentonGentry!

**Comment by DentonGentry** [comment]
Blocked by https://github.com/tailscale/tailscale/issues/1970

**Comment by agrw** [comment]
How is "support MacOS as an exit node" going? Still looking out for this feature.

**Comment by Stickyhavr** [comment]
Me, too. Any update on this? Would love to be able to use the MacOS app as an exit node. Hopefully this will be possible soon. Thanks!

**Comment by DmitriyMV** [comment]
@DentonGentry since #1970 was just fixed are there any other blockers for this?

**Comment by DentonGentry** [comment]
We'll test the functionality during the 1.20 release cycle in January and see what is left to be done.

**Comment by Stickyhavr** [comment]
Ah ok, that's good to know. Still, it will be much nicer once it makes its
way into the App Store or Standalone (GUI) versions. Hopefully work
continues in that direction in the future. Thanks for the update about
Tailscaled.

>

**Comment by bradfitz** [comment]
This is now done. You'll see this done in our private macOS GUI repo. That commit is about to land, which will close this bug. It's expected to be released in our next stable build (1.20.0) after the holidays in early January probably. There might be pre-release "unstable" builds (1.19.x) before then of the DMG or the macOS TestFlight for Monterey TestFlight users.
